### ISSUE TITLE
[Tkinter] OpenGL crash when restoring the window

### ISSUE THREAD

**Issue by ThenTech** [vispy/vispy#1993]
I have a scene containing a `visuals.Plane`, `visuals.XYZAxis` and some `visuals.Text`. If I minimise the window and then restore it, I get errors similar to this one:

```python
[...]
  File "vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "vispy\scene\canvas.py", line 217, in on_draw
    self._draw_scene()
  File "vispy\scene\canvas.py", line 266, in _draw_scene
    self.draw_visual(self.scene)
  File "vispy\scene\canvas.py", line 304, in draw_visual
    node.draw()
  File "vispy\scene\visuals.py", line 99, in draw
    self._visual_superclass.draw(self)
  File "vispy\visuals\visual.py", line 444, in draw
    self._vshare.index_buffer)
  File "vispy\visuals\shaders\program.py", line 101, in draw
    Program.draw(self, *args, **kwargs)
  File "vispy\gloo\program.py", line 535, in draw
    canvas.context.flush_commands()
  File "vispy\gloo\context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "vispy\gloo\glir.py", line 573, in flush
    self._shared.flush(parser)
  File "vispy\gloo\glir.py", line 495, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "vispy\gloo\glir.py", line 820, in parse
    self._parse(command)
  File "vispy\gloo\glir.py", line 780, in _parse
    ob.draw(*args)
  File "vispy\gloo\glir.py", line 1349, in draw
    self._pre_draw()
  File "vispy\gloo\glir.py", line 1283, in _pre_draw
    func(attr_handle, *args)
  File "vispy\gloo\gl\_pyopengl2.py", line 223, in glVertexAttribPointer
    return GL.glVertexAttribPointer(indx, size, type, normalized, stride, offset)
  File "OpenGL\latebind.py", line 63, in __call__
    return self.wrapperFunction( self.baseFunction, *args, **named )
  File "OpenGL\GL\VERSION\GL_2_0.py", line 473, in glVertexAttribPointer
    ArrayDatatype.voidDataPointer( array )
  File "OpenGL\latebind.py", line 43, in __call__
    return self._finalCall( *args, **named )
  File "OpenGL\wrapper.py", line 700, in wrapperCall
    raise err
  File "OpenGL\wrapper.py", line 693, in wrapperCall
    result = wrappedOperation( *cArguments )
  File "OpenGL\error.py", line 234, in glCheckError
    baseOperation = baseOperation,
OpenGL.error.GLError: GLError(
        err = 1281,
        description = b'invalid value',
        baseOperation = glVertexAttribPointer,
        pyArgs = (
                0,
                2,
                GL_FLOAT,
                GL_FALSE,
                8,
                c_void_p(None),
        ),
```

or this one:

```python
OpenGL.error.GLError: GLError(
        err = 1281,
        description = b'invalid value',
        baseOperation = glTexSubImage2D,
        pyArgs = (
                GL_TEXTURE_2D,
                0,
                1050,
                0,
                50,
                100,
                GL_LUMINANCE,
                GL_FLOAT,
                array([[[-0.0947817 ],
        [-0.14582516],
        [ 0.27695575],
        ...,
        [-0.10596392],
        [ 0....,
        ),
```

This is with the `gl+` backend, as the default one just exits the Python process with an error code of 0xC0000005 (Access Violation).
According to the OpenGL reference, the only possible wrong parameter would be the `index`, 0 in this case. Maybe whatever buffer is bound, somehow gets destroyed after restoring the window?

The first shown error is thrown after execution [this call in `_pyopengl2.py`](https://github.com/vispy/vispy/blob/c4f7ef42988dd9312ef7e7a03f4f10347a8e4b07/vispy/gloo/gl/_pyopengl2.py#L223) which was called from [here in glir.py](https://github.com/vispy/vispy/blob/c4f7ef42988dd9312ef7e7a03f4f10347a8e4b07/vispy/gloo/glir.py#L1279).

Not that this:

- *Does not* happen in other backends (as far as I've tested);
- *Does not* happen when using the [pyopengltk frame](https://github.com/jonwright/pyopengltk) on its own (what the VisPy Tkinter backend uses);
- And *does* happen both when embedding the `SceneCanvas` in an existing window or as standalone.

Refer to [my test file](https://gist.github.com/ThenTech/620d2422a8a2f2529d01bf34e229d7ca#file-tk_pyvis-py) that can reproduce the error.

_Originally posted by @ThenTech in https://github.com/vispy/vispy/issues/1918#issuecomment-764965320_

**Comment by Wosser1sProductions** [comment]
Might have a similar underlying cause as #1759 

Output of running my test file with `--vispy-log=debug --vispy-gl-debug` (after restoring the window):
```python
vispy.scene.canvas:SceneCanvas.on_draw(212): DEBUG: Canvas draw
vispy.visuals.shaders.program:ModularProgram.update_variables(145): DEBUG: Apply variables:
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_alpha = 1.0
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_center = (0, 0)
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_enabled = 0
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_id_color = [0.023529411764705882, 0.0, 0.0, 0.0]
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_mag = 1.0
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_matrix = [[ 1.21904762e+02  0.00000000e+00  0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00 -6.57142868e+01  0.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00  9.99999997e-07  0.00000000e+00]
 [ 0.00000000e+00  6.57142883e+02  0.00000000e+00  1.00000000e+00]]
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_radii = (197.14285714285708, 328.5714285714285)
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_scale = [1.    0.125 1.    1.   ]
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_scale_1 = [ 0.0015625  -0.00289855  1.          1.        ]
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_trans = <Texture2D shape=(1000, 1, 1) format='luminance' at 0x21e056e6588>
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_trans_max = 328.5714
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_translate = [0. 0. 0. 0.]
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_translate_1 = [-0.9047619  1.         0.         0.       ]
vispy.visuals.shaders.program:ModularProgram.update_variables(154): DEBUG:     u_view = (61.9523811340332, 1279.0000001816522, 33.857142857143, 689.0)
vispy.visuals.shaders.program:ModularProgram.update_variables(160): DEBUG: Discarding unused variables before draw: dict_keys([])
vispy.gloo.program:ModularProgram.draw(528): DEBUG: Program drawing 'line_strip' with (0, 200000)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glEnable(34370)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glEnable(34913)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glBindFramebuffer(GL_FRAMEBUFFER, 0)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glClearColor(0.0, 0.0, 0.0, 1.0)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glClear(17664)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glDisable(GL_CULL_FACE)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glLineWidth(1.0)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glEnable(GL_DEPTH_TEST)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glEnable(GL_BLEND)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA, GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
vispy.gloo.gl:gl_debug_wrapper(167): DEBUG: glUseProgram(1)
vispy.visuals.visual:ScrollingLines.draw(446): WARNING: Error drawing visual <ScrollingLines at 0x21e02220348>
vispy.util.logs:log_exception(271): WARNING: Traceback (most recent call last):
[...]
  << caught exception here: >>
  File "vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "vispy\scene\canvas.py", line 217, in on_draw
    self._draw_scene()
  File "vispy\scene\canvas.py", line 266, in _draw_scene
    self.draw_visual(self.scene)
  File "vispy\scene\canvas.py", line 304, in draw_visual
    node.draw()
  File "vispy\scene\visuals.py", line 99, in draw
    self._visual_superclass.draw(self)
  File "vispy\visuals\visual.py", line 444, in draw
    self._vshare.index_buffer)
  File "vispy\visuals\shaders\program.py", line 101, in draw
    Program.draw(self, *args, **kwargs)
  File "vispy\gloo\program.py", line 535, in draw
    canvas.context.flush_commands()
  File "vispy\gloo\context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "vispy\gloo\glir.py", line 573, in flush
    self._shared.flush(parser)
  File "vispy\gloo\glir.py", line 495, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "vispy\gloo\glir.py", line 820, in parse
    self._parse(command)
  File "vispy\gloo\glir.py", line 780, in _parse
    ob.draw(*args)
  File "vispy\gloo\glir.py", line 1349, in draw
    self._pre_draw()
  File "vispy\gloo\glir.py", line 1273, in _pre_draw
    self.activate()
  File "vispy\gloo\glir.py", line 1072, in activate
    gl.glUseProgram(self._handle)
  File "vispy\gloo\gl\__init__.py", line 177, in gl_debug_wrapper
    check_error(fn.__name__)
  File "vispy\gloo\gl\__init__.py", line 207, in check_error
    raise err
RuntimeError: OpenGL got errors (glUseProgram): GL_INVALID_VALUE
vispy.util.logs:_handle_exception(325): ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (tkinter) at 0x21e021b5e88>> for DrawEvent
vispy.app.canvas:SceneCanvas.close(465): DEBUG: Closing canvas <SceneCanvas (tkinter) at 0x21e021b5e88>
```

**Comment by djhoese** [comment]
Is it tied to any specific visuals? Like what if you take the `examples/basics/gloo/start.py` and run it with the tk backend?

Is there a chance that your backend has to re-establish the GL context being used? That's my best guess for why you'd be getting these errors. I wonder if there is something special pyopengltk does on minimize/maximize.

**Comment by Wosser1sProductions** [comment]
The basic example you mentioned, `examples/basics/gloo/start.py`, works, but none of the others do (literally any other one I tried).
=> So starting it and looking/interacting is fine, but when restoring after minimizing the window, the same types of errors occur as above.

**Comment by djhoese** [comment]
Not sure it is exactly related but I'm starting to see Tk failures on some of my PRs now: https://github.com/vispy/vispy/pull/1920/checks?check_run_id=1994599145

**Comment by fishymonarch** [comment]
Thios appears to be to an issue with pyopengltk:
https://github.com/jonwright/pyopengltk/issues/29

**Comment by ThenTech** [comment]
@fishymonarch Well what do you know, that was indeed it. I fiddled around with those functions a lot when I was working on my project, but could never fix it. Thanks.

@djhoese I guess this can be closed as it should be resolved in `pyopengltk>=0.0.4`.
